Ccm chunked - enable multipart CCM in PSA

mstarzyk-mobica opened the pull request:
resolves #4353

mstarzyk-mobica commented:
No chunked data support yet. I just need a confirmation that this is correct approach, especially when it comes to the function `mbedtls_ccm_calculate_first_block`

mstarzyk-mobica commented on library/ccm.c:
```diff
@@ -192,23 +246,28 @@ static int ccm_auth_crypt( mbedtls_ccm_context *ctx, int mode, size_t length,
      * 5 .. 3   (t - 2) / 2
      * 2 .. 0   q - 1
      */
-    b[0] = 0;
```
I hoped that `ccm_init()` will take care of clearing the `b[0]`, but it turns out that this has to be cleared inside either
`ccm_starts()` or `ccm_set_lengths()`. A self-test function at the bottom of this file is calling `ccm_init()` only once, and then performing encrypt/decrypt cycle on some test data. This means that the `b[0]` is not cleared between `mbedtls_ccm_encrypt_and_tag` and `mbedtls_ccm_auth_decrypt`

This can be tackled in two ways:
1. Move b[0] calculation to `mbedtls_ccm_calculate_first_block`. This will require storing all necessary variables in the ccm_context
2. Create additional auxiliary function, to conditionally clear the b[0] at the beginning of `ccm_starts` and `ccm_set_lengths` (same mechanism as in line 151)

mstarzyk-mobica replied:
Update - went with second approach

gilles-peskine-arm commented on library/ccm.c:
```diff
@@ -108,82 +106,63 @@ void mbedtls_ccm_free( mbedtls_ccm_context *ctx )
-    if( ( ret = mbedtls_cipher_update( &ctx->cipher_ctx, y, 16, y, &olen ) ) != 0 ) \
-        return( ret );
+#define CCM_STATE__CLEAR                0
+#define CCM_STATE__STARTED              0x0001
+#define CCM_STATE__LENGHTS_SET          0x0002
+#define CCM_STATE__ERROR                0x0004
```
Since these are flags and not alternatives, they might be clearer defined as `(1 << 0)`, `(1 << 1)`, etc.

gilles-peskine-arm commented on library/ccm.c:
```diff
@@ -339,17 +485,25 @@ int mbedtls_ccm_encrypt_and_tag( mbedtls_ccm_context *ctx, size_t length,
 
-    return( mbedtls_ccm_star_encrypt_and_tag( ctx, length, iv, iv_len, add,
-                add_len, input, output, tag, tag_len ) );
+static int mbedtls_ccm_compare_tags(const unsigned char *tag1, const unsigned char *tag2, size_t tag_len)
```
This is for decryption so it should be below the “Authenticated decryption” comment.

gilles-peskine-arm commented on library/ccm.c:
```diff
@@ -363,31 +517,18 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
 {
```
Why did you duplicate the code between `mbedtls_ccm_star_auth_decrypt` and `mbedtls_ccm_auth_decrypt`? They still do the same thing except for passing a different mode to `ccm_auth_crypt`.

mstarzyk-mobica replied:
Passing different mode to ccm_auth_crypt is the only reason. I could add extra function including common code, parametrized with `mode` (basically same arguments as `ccm_auth_crypt`)

gilles-peskine-arm commented on tests/suites/test_suite_ccm.function:
```diff
@@ -1,5 +1,67 @@
+    mbedtls_free( output );
+    output = NULL;
+
+    if( tag->len == 0 )
+        ASSERT_ALLOC( output, 16 );
```
Why? What's special about tag length being 0? If there's no tag then `finish` should be happy with a 0-length buffer, and `ASSERT_ALLOC( output, tag->len );` will do the right thing.

mstarzyk-mobica replied:
ASSERT_ALLOC will not allocate memory for length == 0.
This leads to undefined behaviour.

Although now that I think about it, perhaps it's the `ccm_finish` function that should actually check the provided tag_len https://github.com/ARMmbed/mbedtls/pull/4708/files/9a7a0bf82ccff969ec977918687ee50a839d6b1d#diff-85da8de2cd1010803e0b2fdff14d7af5aa99e1d8d154086e4d6498a1d5afae5fR432

gilles-peskine-arm replied:
Where a length of 0 is expected, a null pointer should work. If there's undefined behavior with pointer=NULL and length=0, this needs to be fixed.

gilles-peskine-arm commented on tests/suites/test_suite_ccm.function:
```diff
@@ -122,71 +184,115 @@ void mbedtls_ccm_encrypt_and_tag( int cipher_id, data_t * key,
 
-    /* Check we didn't write past the end */
-    TEST_ASSERT( msg_n_tag[result->len] == 0 && msg_n_tag[result->len + 1] == 0 );
+    /* Prepare data_t structers for multipart testing */
```
```suggestion
    /* Prepare data_t structures for multipart testing */
```
and another one below

gilles-peskine-arm commented on tests/suites/test_suite_ccm.function:
```diff
@@ -122,71 +184,115 @@ void mbedtls_ccm_encrypt_and_tag( int cipher_id, data_t * key,
+    /* Prepare data_t structers for multipart testing */
+    const data_t encrypted_expected = { .x = result->x,
+                                        .len = msg->len };
+    const data_t tag_expected = { .x = (uint8_t*) expected_tag, /* cast to conform with data_t x type */
```
Rather than cast, how about constructing this structure first at the top of the function, and then optionally defining `const * expected_tag` from it/

mstarzyk-mobica replied:
I'd rather keep the cast, rather then reorganize the function, as that would reduce readability.
I can get rid of the const attribute from the `expected_tag` variable.

gilles-peskine-arm commented on tests/suites/test_suite_ccm.function:
```diff
@@ -1,5 +1,67 @@
 /* BEGIN_HEADER */
```
Missing: tests that skip calling `update_ad` when the AD is empty, that skip calling `update` when the plaintext/ciphertext is empty, or both.

mstarzyk-mobica commented on library/ccm.c:
```diff
@@ -254,62 +228,240 @@ static int ccm_auth_crypt( mbedtls_ccm_context *ctx, int mode, size_t length,
+                return ret;
+
+            for( i = 0; i < use_len; i++ )
+                ctx->y[i + offset] ^= output[i];
```
Actually I was using the gcm.c as reference.
Have a look at https://github.com/ARMmbed/mbedtls/blob/development/library/gcm.c#L440
This function is later called by:
* https://github.com/ARMmbed/mbedtls/blob/development/library/gcm.c#L495
* https://github.com/ARMmbed/mbedtls/blob/development/library/gcm.c#L512
The `out_p` pointer is `output` https://github.com/ARMmbed/mbedtls/blob/development/library/gcm.c#L452

mstarzyk-mobica replied:
Does this mean the gcm code needs to be fixed as well?

gilles-peskine-arm replied:
Yes, GCM needs to be fixed as well. But this is preexisting in CCM, so it's ok not to fix it in this PR.

mstarzyk-mobica commented on library/ccm.c:
```diff
@@ -192,58 +171,53 @@ static int ccm_auth_crypt( mbedtls_ccm_context *ctx, int mode, size_t length,
-            len_left -= use_len;
-            src += use_len;
-        }
+    if( ctx->state & CCM_STATE__ERROR )
+    {
+        mbedtls_ccm_clear_state(ctx);
     }
```
My assumption was that in case of error during `ccm_update` or `update_ad`, user might restart the crypting operation without calling `ccm_finish`. If we know the operation failed during, for example, `ccm_update`, what is the point of calling `ccm_finish`?

gilles-peskine-arm replied:
Reusing an operation object after an error is not very useful, and in any case it can only work if you know the state of the object and that state is useful. Here the state on error isn't documented, but remains an error state (which is good: it avoids accidental misuse if the user doesn't check the return status of an intermediate function).

mstarzyk-mobica replied:
Alright, so we assume the user will reinit the context if the error occured.
Should I add a check for error state at the beginning of the new (chunked data) functions?

mstarzyk-mobica commented on library/ccm.c:
```diff
@@ -254,62 +228,240 @@ static int ccm_auth_crypt( mbedtls_ccm_context *ctx, int mode, size_t length,
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    unsigned char i;
+    size_t olen, use_len, offset;
+
```
- [ ] Add tests for input data lengths

mstarzyk-mobica commented on library/ccm.c:
```diff
@@ -254,62 +228,240 @@ static int ccm_auth_crypt( mbedtls_ccm_context *ctx, int mode, size_t length,
+    }
+
+    if( ctx->processed - 2 == ctx->add_len )
+        ctx->processed = 0; // prepare for mbedtls_ccm_update()
```
This test scenario is covered by multipart function: when `n1_add` = add_len, then `n2_add` = 0. Same for plaintext and `n1`

gilles-peskine-arm commented on tests/suites/test_suite_ccm.function:
```diff
@@ -457,3 +454,114 @@ exit:
+/* END_CASE */
+
+/* BEGIN_CASE */
+void mbedtls_ccm_skip_cipher( int cipher_id, int mode,
```
What does “skip cipher” mean? Maybe this should be `mbedtls_ccm_skip_update`?

gilles-peskine-arm commented on tests/suites/test_suite_ccm.function:
```diff
@@ -457,3 +454,114 @@ exit:
 /* END_CASE */
+
+/* BEGIN_CASE */
+void mbedtls_ccm_skip_auth( int cipher_id, int mode,
```
“skip auth” makes me think this is about the tag. I think this should be `mbedtls_ccm_skip_ad`.

gilles-peskine-arm commented on tests/suites/test_suite_ccm.function:
```diff
@@ -457,3 +454,114 @@ exit:
+/* END_CASE */
+
+/* BEGIN_CASE */
+void mbedtls_ccm_overflow_auth( int cipher_id, int mode,
```
As above:  `mbedtls_ccm_overflow_ad`

gilles-peskine-arm commented on tests/suites/test_suite_ccm.function:
```diff
@@ -457,3 +454,114 @@ exit:
+/* END_CASE */
+
+/* BEGIN_CASE */
+void mbedtls_ccm_overflow_cipher( int cipher_id, int mode,
```
As above: `mbedtls_ccm_overflow_input` or `mbedtls_ccm_overflow_data` maybe (could be ciphertext or plaintext)

gilles-peskine-arm commented on tests/suites/test_suite_ccm.function:
```diff
@@ -457,3 +454,114 @@ exit:
+    // set half of auth data length to provoke an overflow
+    TEST_EQUAL( 0, mbedtls_ccm_set_lengths( &ctx, add->len / 2, 16, 16 ) );
+
+    TEST_EQUAL( MBEDTLS_ERR_CCM_BAD_INPUT, mbedtls_ccm_update_ad( &ctx, add->x, add->len) );
```
There should also be a test with a first call to update_ad that fills the expected length, and another call that tries to go beyond. The logic is different for those two cases. 

Likewise with update.

It would be worth looking at the test coverage and ensuring that all code paths are covered except maybe for some hard-to-trigger error cases (e.g. failure of the underlying cipher).

gilles-peskine-arm commented on tests/suites/test_suite_ccm.function:
```diff
@@ -249,45 +386,182 @@ void mbedtls_ccm_star_auth_decrypt( int cipher_id,
+    mbedtls_free( output );
     mbedtls_ccm_free( &ctx );
 }
 /* END_CASE */
```
Please add test cases for calling finish with insufficient AD, or with insufficient plaintext/ciphertext.

gilles-peskine-arm replied:
Please add test cases that just call finish, i.e. empty AD and empty plaintext/ciphertext.

mstarzyk-mobica commented on library/ccm.c:
```diff
@@ -263,39 +246,51 @@ int mbedtls_ccm_set_lengths( mbedtls_ccm_context *ctx,
-    if( ctx->add_len > 0 && add_len > 0)
+    if( ctx->state & CCM_STATE__ERROR )
     {
+        return ret;
```
Actually CCM_STATE__ERROR is set when any other part of the library used in ccm returns an error. So the error CORRUPTION_DETECTED is quite fitting here. However, there are other places in which it is misused, I will fix them

mstarzyk-mobica replied:
Also - MBEDTLS_ERR_CCM_BAD_INPUT seems to be a suitable error to return in case of invalid lengths.

mstarzyk-mobica commented on library/ccm.c:
```diff
@@ -263,39 +246,51 @@ int mbedtls_ccm_set_lengths( mbedtls_ccm_context *ctx,
+    {
+        if( ctx->state & CCM_STATE__AUTH_DATA_FINISHED )
+        {
+            return ret;
```
> why error out on extra additional data only when ctx->add_len > 0?

Actually the `ctx->add_len > 0` is not really related to the `ctx->state & CCM_STATE__AUTH_DATA_FINISHED`.
The `ctx->add_len > 0` is there to ignore any input if configured auth data length was zero.

As for the error in case of `ctx->state & CCM_STATE__AUTH_DATA_FINISHED` (auth data exceeded set length), it is dependant on the `add_len` rather then `ctx->add_len`. We do not want the function to return an error in following scenario:
1. Pass all auth data to ccm_update_ad
2. Call ccm_update_ad with empty input (zero length)

mstarzyk-mobica commented:
Rebased on development to fix conflict on ccm.h (occured after moving trailing comments for error definitions)
Added missing tests.
Implemented requested changes.

lukgni commented on ChangeLog.d/chunked_ccm.txt:
```diff
@@ -0,0 +1,8 @@
+     The multi-part functions: mbedtls_ccm_starts(), mbedtls_ccm_set_lengths(),
+     mbedtls_ccm_update_ad(), mbedtls_ccm_update(), mbedtls_ccm_finish()
+     were introduced in mbedTLS 3.0 release, however their implementation was
+     postponed util now.
```
util -> until

mstarzyk-mobica replied:
fixed

lukgni reviewed (changes requested):
One small typo - besides that LGTM

mstarzyk-mobica commented:
Conflict resolved by merging `development` into current branch.

ronald-cron-arm commented:
> Conflict resolved by merging `development` into current branch.

Please rebase instead as rebasing is the preferred way to solve PR conflicts in the project.

mstarzyk-mobica commented:
Rebasing would require fixing conflicts in almost individual conflict. I am aware that rebase is prefered to merge, however in this case merge required significantly less effort.

lukgni reviewed (dismissed):
LGTM!

gilles-peskine-arm commented on library/ccm.c:
```diff
@@ -253,62 +211,290 @@ static int ccm_auth_crypt( mbedtls_ccm_context *ctx, int mode, size_t length,
-            if( ++ctr[15-i] != 0 )
+    if( ctx->state & CCM_STATE__ERROR )
+    {
+        return MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
```
This is more likely to be BAD_SEQUENCE (application error) than CORRUPTION_DETECTED (system error).

gilles-peskine-arm commented on include/mbedtls/ccm.h:
```diff
@@ -76,7 +78,27 @@ extern "C" {
+    size_t MBEDTLS_PRIVATE(tag_len);         /*!< Total tag length */
+    size_t MBEDTLS_PRIVATE(processed);       /*!< Track how many bytes of input data
+                                                  were processed (chunked input).
+                                                  Used indepenedantly for both auth data
```
Minor: indepenedantly → independently

gilles-peskine-arm commented on tests/suites/test_suite_ccm.data:
```diff
@@ -1554,30 +1554,94 @@ CCM encrypt, overflow ad NIST VADT AES-192 #14 (P=24, N=13, A=13, T=16)
 depends_on:MBEDTLS_AES_C
 mbedtls_ccm_overflow_ad:MBEDTLS_CIPHER_ID_AES:MBEDTLS_CCM_ENCRYPT:"b5f43f3ae38a6165f0f990abe9ee50cd9ad7e847a0a51731":"13501aebda19a9bf1b5ffaa42a":"ead4c45ff9db54f9902a6de181"
 
+CCM encrypt, incomplete ad NIST VADT AES-192 #14 (P=24, N=13, A=13, T=16)
```
Please change the corner case and instant finish test cases to using AES-128. Some platforms lack AES-192, but they should still be able to run the corner case tests.

gilles-peskine-arm commented on tests/suites/test_suite_ccm.function:
```diff
@@ -539,6 +539,53 @@ exit:
+    // use hardcoded values for msg length and tag length. They are not a part of this test
+    TEST_EQUAL( 0, mbedtls_ccm_set_lengths( &ctx, add->len, 0, 16 ) );
+
+    TEST_EQUAL( 0, mbedtls_ccm_update_ad( &ctx, add->x, add->len/2) );
```
Minor: `add_len - 1` would be better validation: it would give more confidence that the code is checking the correct value and isn't off by one.

Same with update.

gilles-peskine-arm commented on tests/suites/test_suite_ccm.function:
```diff
@@ -539,6 +539,53 @@ exit:
+    // pass full auth data
+    TEST_EQUAL( 0, mbedtls_ccm_update_ad( &ctx, add->x, add->len) );
+    // pass 1 extra byte
+    TEST_EQUAL( MBEDTLS_ERR_CCM_BAD_SEQUENCE, mbedtls_ccm_update_ad( &ctx, add->x, 1) );
```
This tests one scenario: passing exactly the required amount of data, then passing more data. There should also be tests for other scenarios which lead to different code paths: passing too much data the first time; passing partial data then jumping over the required amount (e.g. pass n-1 then 2).

Same with update.

mstarzyk-mobica replied:
"passing too much data the first time" is covered by test `mbedtls_ccm_overflow_ad`
"passing partial data then jumping over the required amount (e.g. pass n-1 then 2)." <- this is indeed missing

gilles-peskine-arm commented on tests/suites/test_suite_ccm.data:
```diff
@@ -1517,3 +1517,143 @@ mbedtls_ccm_encrypt_and_tag:MBEDTLS_CIPHER_ID_CAMELLIA:"D75C2778078CA93D971F96FD
+CCM* decrypt, full update and overflow NIST VADT AES-192 #14 (P=24, N=13, A=13, T=16)
+depends_on:MBEDTLS_AES_C
+mbedtls_ccm_full_update_and_overflow:MBEDTLS_CIPHER_ID_AES:MBEDTLS_CCM_STAR_DECRYPT:"b5f43f3ae38a6165f0f990abe9ee50cd9ad7e847a0a51731":"3726c1aaf85ee8099a7ebd3268700e07d4b3f292c65bba34":"13501aebda19a9bf1b5ffaa42a":"ead4c45ff9db54f9902a6de181"
+
```
Missing good cases: not calling `update_ad` at all (with A=0); not calling either `update_ad` or `update` (with A=0 and P=0). I.e. similar to `mbedtls_ccm_skip_update`, but skipping `update_ad`, or skipping both.

mstarzyk-mobica replied:
Please be more specific here. Calling `update` and `update_ad` with 0 input is covered by `check_multipart` function, `n1` and `n1_add` are iterated from 0 to `n`.

gilles-peskine-arm replied:
What's covered is calling `update_ad(0); update(0);`. But **not calling the functions** is not covered. We need to cover all of the following call sequences (where `a`, `p` are `> 0`):
```
{
    start();
    set_lengths(a, p);
    update_ad(a);
    update(p);
    finish();
}

{
    start();
    set_lengths(0, p);
    update_ad(0);
    update(p);
    finish();
}

{
    start();
    set_lengths(0, p);
    // no update_ad
    update(p);
    finish();
}

{
    start();
    set_lengths(a, 0);
    update_ad(a);
    update(0);
    finish();
}

{
    start();
    set_lengths(a, 0);
    update_ad(a);
    // no update
    finish();
}

{
    start();
    set_lengths(0, 0);
    update_ad(0);
    update(0);
    finish();
}

{
    start();
    set_lengths(0, 0);
    // no update_ad
    update(0);
    finish();
}

{
    start();
    set_lengths(0, 0);
    update_ad(0);
    // no update
    finish();
}

{
    start();
    set_lengths(0, 0);
    // no update_ad
    // no update
    finish();
}
```

This can be done with separate functions or by leveraging the existing data and adding additional calls to `check_multipart`which skip the `update_ad` call, skip the `update` call, or both.

mstarzyk-mobica replied:
```
{
    start();
    set_lengths(a, p);
    update_ad(a);
    update(p);
    finish();
} - covered by mbedtls_ccm_encrypt_and_tag and mbedtls_ccm_auth_decrypt tests, such as:
CCM encrypt and tag NIST VNT AES-192 #1 (P=24, N=7, A=32, T=16)
CCM auth decrypt tag NIST DVPT AES-128 #27 (P=24, N=7, A=32, T=16)

{
    start();
    set_lengths(0, p);
    update_ad(0);
    update(p);
    finish();
} - covered by: mbedtls_ccm_encrypt_and_tag and mbedtls_ccm_auth_decrypt tests, such as:
CCM encrypt and tag NIST VADT AES-128 #1 (P=24, N=13, A=0, T=16)
CCM auth decrypt tag NIST DVPT AES-128 #11 (P=24, N=7, A=0, T=16)

{
    start();
    set_lengths(0, p);
    // no update_ad
    update(p);
    finish();
} - covered by mbedtls_ccm_skip_ad, such as:
CCM encrypt, skip ad NIST VADT AES-128 (P=24, N=13, A=0, T=16)
CCM decrypt, skip ad NIST DVPT AES-192 (P=24, N=7, A=0, T=4)

{
    start();
    set_lengths(a, 0);
    update_ad(a);
    update(0);
    finish();
} - covered by: mbedtls_ccm_encrypt_and_tag and mbedtls_ccm_auth_decrypt tests, such as:
CCM encrypt and tag NIST VPT AES-128 #1 (P=0, N=13, A=32, T=16)
CCM auth decrypt tag NIST DVPT AES-128 #17 (P=0, N=7, A=32, T=4)

{
    start();
    set_lengths(a, 0);
    update_ad(a);
    // no update
    finish();
} - covered by mbedtls_ccm_skip_update tests, such as:
CCM encrypt, skip update NIST VPT AES-128 #1 (P=0, N=13, A=32, T=16)
CCM decrypt, skip update NIST DVPT AES-256 #23 (P=0, N=13, A=32, T=16)

{
    start();
    set_lengths(0, 0);
    update_ad(0);
    update(0);
    finish();
} - covered by mbedtls_ccm_auth_decrypt tests for P=0, A=0
such as: CCM auth decrypt tag NIST DVPT AES-128 #1 (P=0, N=7, A=0, T=4)
Missing: mbedtls_ccm_encrypt_and_tag tests for P=0, A=0

{
    start();
    set_lengths(0, 0);
    // no update_ad
    update(0);
    finish();
} - missing mbedtls_ccm_skip_ad tests for P=0, A=0

{
    start();
    set_lengths(0, 0);
    update_ad(0);
    // no update
    finish();
} - missing mbedtls_ccm_skip_update tests for P=0, A=0

{
    start();
    set_lengths(0, 0);
    // no update_ad
    // no update
    finish();
} - missing mbedtls_ccm_instant_finish P=0, A=0
```
so:
- [ ] add mbedtls_ccm_encrypt_and_tag tests for P=0, A=0
- [ ] add mbedtls_ccm_skip_ad tests for P=0, A=0
- [ ] add mbedtls_ccm_skip_update tests for P=0, A=0
- [ ] add mbedtls_ccm_instant_finish P=0, A=0

mstarzyk-mobica replied:
Can't find [here](https://csrc.nist.gov/projects/cryptographic-algorithm-validation-program/cavp-testing-block-cipher-modes#CCM) a proper test vector (P=0, A=0) for mbedtls_ccm_encrypt_and_tag test

gilles-peskine-arm replied:
If you can't find a test vector matching some conditions from NIST or an RFC, you can build one with a well-established implementation such as python/cryptodome or openssl.

gilles-peskine-arm commented on library/ccm.c:
```diff
@@ -253,62 +211,290 @@ static int ccm_auth_crypt( mbedtls_ccm_context *ctx, int mode, size_t length,
+        return ret;
+    }
+
+    if( ctx->add_len > 0 && add_len > 0 )
```
(Continued from https://github.com/ARMmbed/mbedtls/pull/4708#discussion_r684254230)

If `ctx->add_len == 0` and `add_len == 0`, i.e. if empty AD is expected but the application passes non-empty AD, this function should return `MBEDTLS_ERR_CCM_BAD_INPUT`, but this isn't happening.

I think removing `ctx->add_len > 0 &&` would fix that.

Please add a non-regression test.

gilles-peskine-arm reviewed (changes requested):
Looks mostly good, but there are still a few missing tests for corner cases and a couple of issues with detection of bad usage.

mstarzyk-mobica commented on include/mbedtls/ccm.h:
```diff
@@ -61,6 +61,8 @@
 #define MBEDTLS_ERR_CCM_BAD_INPUT       -0x000D
 /** Authenticated decryption failed. */
 #define MBEDTLS_ERR_CCM_AUTH_FAILED     -0x000F
+/** CCM functions called in the wrong sequence. */
```
In one of previous reviews you said "I think we need a new error code here for calling CCM functions out of sequence or with lengths that don't match what was passed to set_lengths()."
Are you suggesting that I should remove the BAD_SEQUENCE error and use only BAD_INPUT?

gilles-peskine-arm replied:
I think using BAD_INPUT for something like `update(); update_ad();` is a bit of a stretch, which is why I suggested BAD_SEQUENCE. On the other hand, the API functions don't always have enough information to distinguish between e.g. calling a function at the wrong time and passing too much input, which means that sometimes the choice of BAD_SEQUENCE vs BAD_INPUT can be misleading. So I'm ok with a single error code, and I'm ok with two error codes, but if there are two error codes, the code and the documentation need to be made clearer so that the distinction between these two errors is easy to understand. I think it's easier to stick to a single error code.

mstarzyk-mobica commented:
Second reviewer still needed.
Lukasz left but he approved the changes up to `e40ae6b`, so the review can be continued from commit `e40ae6b`

gilles-peskine-arm commented on tests/suites/test_suite_ccm.function:
```diff
@@ -591,6 +591,41 @@ exit:
+     * with added extra byte at the end */
+    uint8_t* add_extended = NULL;
+    ASSERT_ALLOC( add_extended, add->len + 1 );
+    if( add_extended )
```
This branch is always taken since the previous line either allocates a nonzero size or jumps to `exit`.

Also goes for `mbedtls_ccm_incomplete_update_overflow`.

gilles-peskine-arm commented on tests/suites/test_suite_ccm.function:
```diff
@@ -591,6 +591,41 @@ exit:
+    /* New auth buffer containing same data as original one,
+     * with added extra byte at the end */
+    uint8_t* add_extended = NULL;
+    ASSERT_ALLOC( add_extended, add->len + 1 );
```
I think it would be simpler to pass `add->len - 1` bytes from `add` first, then pass 2 bytes from a local 2-byte array.

Also goes for `mbedtls_ccm_incomplete_update_overflow`.

gilles-peskine-arm reviewed (changes requested):
There's a little bit of dead code in a test. Other than that, LGTM.

gilles-peskine-arm reviewed (approved):
LGTM
